### ISSUE TITLE
Ass subtitle with local font crashes Kodi

### ISSUE THREAD

**Issue by scott967** [xbmc/xbmc#20108]
### Describe the bug
Here is a clear and concise description of what the problem is:

This is an investigation of user support problem reported here: [https://forum.kodi.tv/showthread.php?tid=364289](url)

To reproduce the problem I tested using Kodi 20 alpha nightly from Aug 15.  Intent was to test with PR19535 [https://github.com/xbmc/xbmc/pull/19535](url) from @enen92 merged.  IIUC with this commit, an external ASS format subtitle file with a style using a font will render if the font is placed in special://home/media/Fonts.  It is unclear from the documentation of the PR or in libass if the font name (internal to ttf font) or filename is required in the ass subtitle style definition.  This test install is on Windows using a portable install and the font was therefor placed in Kodi\portable_data\media\Fonts.  What testing shows is that when the font is installed as a Windows font, or placed in the Kodi\media\Fonts folder (i.e., with arial.ttf) it is not found and instead arial font is used (in this test for the original issue, CJK characters were to be rendered, which arial.ttf does not provide glyphs for).  Because the user issue was with a font  name / filename in CJK characters, I tested with that and also editing the font to use pure ascii in the font name and also filename.  That didn't make any difference (Kodi arial font used to render subtitles).

I then tried creating Kodi\portable_data\media\Fonts and adding the font there (filename and font name set to  DFFangSongW2).  With the ttf font file added, starting the movie playback crashed Kodi.  Removing the file resulted in playback as before.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

When an external subtitle in ass format is provided and the style calls for a font, user can place the appropriate ttf font file into local media\Fonts and libass should find the file and extract the required glyphs.  

## Actual Behavior

Crash after starting playback

## Possible Fix

Review PR19535 for possible errors.  Also might help to document if the font name or filename (or both or font name must be the same as filename) is required in the ass file and the ttf file.

### To Reproduce

1. Add movie to library (tested: Blender "Sintel" test as iso).
2. Add external sub to movie folder (tested: Sintel_zh.ass) 
2. Create local folder media\Fonts (tested: Kodi\portable_data\media\Fonts
3. Add local ttf font (tested: DFFangSongW2.ttf)
4. From Videos window, start playback with ass subtitle selected and subtitles ON (tested: Sintel)
5.  Note this is on Windows with English system UI and en-US locale, Kodi UI language English(US)  Ass subtitle is utf-8 encoded

### Debuglog

The debuglog can be found here:
This debug log is without adding the local font file: [https://paste.kodi.tv/zokegabaki.kodi](url)
Note area of interest at line 2932.  I see there is some "gibberish" in line 2934 (selecting one more font for ( ¶=¾ӂ, 400, 0))
This debug log shows the crash after adding the local font file:  [https://paste.kodi.tv/uqezubozem.kodi](url)

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Here is the subtitle file:  [https://paste.kodi.tv/bujuyitixu](url)  The font file is large (due to coverage for CJK) but can be provided.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10 x64 21H1
 - Kodi version: 20A1 master nightly 2021-08-15 Win x64 build
 

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by enen92** [comment]
@scott967 can you please share your ass file and font?

**Comment by CastagnaIT** [comment]
@enen92 font & subs here: 
[Issue-subtitles-font.zip](https://github.com/xbmc/xbmc/files/7122724/Issue-subtitles-font.zip)

i have tried also another similar font: 
[FangSong_17042.zip](https://github.com/xbmc/xbmc/files/7122729/FangSong_17042.zip)
same crash

by debugging the crash seem happen on libass library
on our implementation of direct write font loading on this line:
https://github.com/CastagnaIT/libass/blob/kodi/libass/ass_directwrite.c#L180
the crash not explains too much:
`Exception thrown at 0x00007FFDF808E451 (DWrite.dll) in kodi.exe: 0xC0000005: Access violation writing location 0x0000000000000000.
`
the filePath appear correct

**Comment by enen92** [comment]
Note the version of libass we use in Kodi for windows has a few patches on top mainly for filepath handling

**Comment by CastagnaIT** [comment]
yes indeed (in my linked gh libass branch i have merged all kodi patches in one commit)

here there are two problems,
the first easy, the ASS file provided have a wrong font name specified in the ASS styles
is a casuality if libass load the right font without fix the file...

in the ASS file there is the font file name:
`Style: Default,華康仿宋體-W2,18,&H00FFFFFF,&H0300FFFF,&H00000000,&H02000000,0,0,0,0,100,100,0,0,1,1,0,2,10,10,10,1`

instead must reflect the real TTF font name:
`Style: Default,DFFangSongW2-B5,18,&H00FFFFFF,&H0300FFFF,&H00000000,&H02000000,0,0,0,0,100,100,0,0,1,1,0,2,10,10,10,1`

after fixed the font name in the ASS file, the font is correctly loaded,
you can test by placing the TTF in the kodi folder /media/fonts (not user folder)

i still not understand well the crash due to loading font in the user folder,
but i have found another thing, seem the fonts placed in the user data folder,
should be loaded from this method:
https://github.com/CastagnaIT/libass/blob/kodi/libass/ass_fontselect.c#L215-L270
but this code have a bug then is not able to scan the user font folder
i have fixed this code but the dwrite crash still happen

**Comment by enen92** [comment]
Just tested your sample on linux and all works great, this might indeed be a bug that is only reproduceable on windows with directwrite

**Comment by CastagnaIT** [comment]
i have fixed the crash in the font loading for user fonts folder
updated commit: https://github.com/CastagnaIT/libass/commits/font_load

but now i have found another error,
if you add these fonts in the kodi/media/fonts (and leave user fonts folder empty)
and you enable ass font override, cause font loading error
only on these fonts (arial/teletext instead works)

**Comment by enen92** [comment]
Nice one! Can you PR libass patches to https://github.com/Paxxi/libass?

I am sure @Paxxi can review them and push a fix to the mirrors

**Comment by CastagnaIT** [comment]
if i am finding to fix the last bug
i would try include also this last fix before sending the changes to paxxi

**Comment by CastagnaIT** [comment]
@scott967 if you can do a confirmation test
just replace the file libass.dll contained in the Kodi directory, with the file within the zip:
[libass_patched.zip](https://github.com/xbmc/xbmc/files/7149179/libass_patched.zip)

**Comment by Paxxi** [comment]
I can't upload to mirrors. I have approved the PR and a build is underway here https://github.com/xbmc/kodi-deps/actions . Once it's done an admin need to fetch the files from there and upload to ftp.

**Comment by CastagnaIT** [comment]
@Paxxi have you check for what i wrote in PR "sideonote" about the Stop crash?
just to make sure you don't update with a library that doesn't work

**Comment by scott967** [comment]
I will test your test build later today, but I have been doing a lot of testing.  I used 1i9.1 release build for tests.  I found that libass is working with DirectWrite (DWrite.dll) backend correctly in almost all configurations.  What I found is that libass will find a font either in installed windows fonts (preferred) or user's media/Fonts.  But there is a bug which I think is maybe in libass.  If the filename in user's media/Fonts folder contains unicode (i.e., not latin-1 0-255) chars in the file name, libass will find the font and select it in font_select, but when it tries to actually open it (I assume to pull glyphs to cache) it will log an error unable to open file.

So I think I understand how libass on windows in 19.1 works (but there needs to be a wiki comment that libass uses the font family name from the ass file to find a compatible font, doesn't care about the font filename in windows). 

But my testing of Kodi 20 Aug 31 win X64 build is giving me all kinds of problems including the crash.  But I will test your build for now.

**Comment by scott967** [comment]
OK I installed your patched libass.dll, and that cured the immediate crash issue and did load what looks like the windows font I installed: `CDVDSubtitlesLibass: [ass] fontselect: (DFFangSongW2-B5, 400, 0) -> DFFangSong-W2-WIN-BF, 0, DFFangSong-W2-WIN-BF`

For this test I have the required font family installed in windows  and a font from a different font family in media/Fonts, so libass did find the windows font family.  But I noticed that in Kodi 19.1, I could see (using sysinternals procmon utility) libass would open every system font but it isn't doing that now.  That's just an observation.  Since "arial" is hard coded in Kodi DVDSubtitlesLibass.cpp as the fallback font family I will test that too (works in Kodi 19.1, libass finds the windows installed arial font and also DWrite will do font substitution for a missing glyph in arial).

So in summary
1.  No crash when windows font family exists and an unrelated font located in user media/Fonts
2. I see different behavior in how libass DirectWrite provider is querying system fonts (don't know if this is a problem)
3.  I will conduct further tests with other configurations

**Comment by scott967** [comment]
OK so tested various configurations, with the font family in windows fonts only, in media/Fonts only, is both, and in neither.  All worked, at least for the ass file / font family I tested (don't know if any impact on stream font from mkv file for example).

Also, now media/Fonts with unicode in filename work, so that's an improvement on Kodi 19.1.

So, this appears to be a good fix.

**Comment by CastagnaIT** [comment]
thanks for your confirmations

If you want try use also this last libass update:
[libass_build2.zip](https://github.com/xbmc/xbmc/files/7189445/libass_build2.zip)

> Since "arial" is hard coded in Kodi DVDSubtitlesLibass.cpp

the font fallback behaviour a bit more complex
the font family set in this case "arial" is a first "suggestion" on which the libass code makes a search
but after this the fallback code behaviour could do also others things
however your explaination is more or less right on Kodi<=19.x the "arial" family fallback should almost always be respected,
and the user cannot change this

but on Kodi 20 + the new Libass implementation: https://github.com/xbmc/xbmc/pull/20010
this has been changed,
we set a fake family name (that could also be left empty), because this take precedence over the our specified font path (font set on kodi subtitle settings)
theoretically (but this depends also on the font provider) if it does not find an approximate font compatibility
should set the font set on kodi subtitle settings

**Comment by enen92** [comment]
The libass packages have been uploaded to the mirrors: https://mirrors.kodi.tv/build-deps/win32/
